tools/windows-node-installer/pkg/cloudprovider/aws/ec2.go

```diff
@@ -606,53 +641,87 @@ func (a *AwsProvider) createWindowsWorkerSg(infraID string, vpc *ec2.Vpc) (strin
-	myIP, err := GetMyIp()
-	if err != nil {
-		return err
+// GetPortsFoSg returns a slice of ports which need to be updated in the rules for local IP
+func getPortsForSg(myIP string, rules []*ec2.IpPermission) []int64 {
+	var ports []int64
+	portTracker := map[int64]bool{}
+	// Loop through the rules, then loop through the ips to get a map of ports
+	if rules != nil {
+		for _, rule := range rules {
+			for _, ips := range rule.IpRanges {
+				if *ips.CidrIp == myIP+"/32" {
+					portTracker[*rule.FromPort] = true
+				}
+			}
+		}
 	}
 
-	_, err = a.EC2.AuthorizeSecurityGroupIngress(&ec2.AuthorizeSecurityGroupIngressInput{
-		GroupId: aws.String(sgID),
-		IpPermissions: []*ec2.IpPermission{
+	// Add ports that need to be updated in a slice ports
+	if !portTracker[WINRM_PORT] {
+		ports = append(ports, WINRM_PORT)
+	}
+	if !portTracker[sshPort] {
+		ports = append(ports, sshPort)
+	}
+	if !portTracker[rdpPort] {
+		ports = append(ports, rdpPort)
+	}
+	return ports
+}
+
+// GetRulesForSgUpdate returns a slice of rules which are required to be updated in sg with local IP
+// This serves as an input to addIngressRules function. This function returns error if failed.
+func (a *AwsProvider) GetRulesForSgUpdate(myIP string, rules []*ec2.IpPermission, vpcCidr string) ([]*ec2.IpPermission, error) {
+	var hasClusterCidrRule bool
+	rulesForUpdate := make([]*ec2.IpPermission, 0)
+
+	ports := getPortsForSg(myIP, rules)
+
+	if rules != nil {
+		// check if cluster communication rule is missing
+		for _, rule := range rules {
+			for _, ips := range rule.IpRanges {
+				if *ips.CidrIp == vpcCidr && *rule.IpProtocol == "-1" {
+					hasClusterCidrRule = true
+					break
+				}
+			}
+		}
+	}
+
+	// If it is missing, add default rule with IpProtocol value "-1"
+	if !hasClusterCidrRule {
+		rulesForUpdate = append(rulesForUpdate,
 			(&ec2.IpPermission{}).
 				// -1 is to allow all ports.
 				SetIpProtocol("-1").
 				SetIpRanges([]*ec2.IpRange{
 					(&ec2.IpRange{}).
-						SetCidrIp(*vpc.CidrBlock),
-				}),
-			(&ec2.IpPermission{}).
-				SetIpProtocol("tcp").
-				SetFromPort(3389).
-				SetToPort(3389).
-				SetIpRanges([]*ec2.IpRange{
-					(&ec2.IpRange{}).
-						SetCidrIp(myIP + "/32"),
-				}),
-			(&ec2.IpPermission{}).
-				SetIpProtocol("tcp").
-				// winrm ansible https port
-				SetFromPort(WINRM_PORT).
-				SetToPort(WINRM_PORT).
-				SetIpRanges([]*ec2.IpRange{
-					(&ec2.IpRange{}).
-						SetCidrIp(myIP + "/32"),
-				}),
-			(&ec2.IpPermission{}).
-				SetIpProtocol("tcp").
-				// SSH port to be opened for
-				// TODO: add validation to check if the port is already opened, inform the user
-				// of the same and this should be a no-op
-				SetFromPort(sshPort).
-				SetToPort(sshPort).
-				SetIpRanges([]*ec2.IpRange{
-					(&ec2.IpRange{}).
-						SetCidrIp(myIP + "/32"),
-				}),
-		},
+						SetCidrIp(vpcCidr),
+				}))
+	}
+
+	// create a list of rules to be updated in sg
+	for _, port := range ports {
```

Comment:
We should actually find a way to generalize the code in `createNewRulesForSg` so that we dont have duplicate code here.